alf/utils/common.py

```diff
@@ -414,30 +414,41 @@ def get_global_counter(default_counter=None):
 
 
 @gin.configurable
-def image_scale_transformer(observation, min=-1.0, max=1.0):
+def image_scale_transformer(observation, fields=None, min=-1.0, max=1.0):
     """Scale image to min and max (0->min, 255->max)
 
-    Note: it treats an observation with len(shape)==4 as image
     Args:
-        observation (nested Tensor): observations
+        observation (nested Tensor): If observation is a nested structure, only
+            namedtuple is supported for now.
+        fields (list[str]): the fields to be applied with the transformation. If
+            None, then `observation` must be a tf.Tensor with dtype uint8. A
+            field str can be a multi-step path denoted by "A.B.C".
         min (float): normalize minimum to this value
         max (float): normalize maximum to this value
     Returns:
         Transfromed observation
     """
 
     def _transform_image(obs):
-        # tf_agent changes all gym.spaces.Box observation to tf.float32.
-        # See _spec_from_gym_space() in tf_agents/environments/gym_wrapper.py
-        # TODO: this shape check is not a good one!
-        if len(obs.shape) >= 4:
-            if obs.dtype == tf.uint8:
-                obs = tf.cast(obs, tf.float32)
-            return ((max - min) / 255.) * obs + min
-        else:
-            return obs
-
-    return tf.nest.map_structure(_transform_image, observation)
+        assert isinstance(obs, tf.Tensor)
+        assert obs.dtype == tf.uint8, "Image must have dtype uint8!"
+        obs = tf.cast(obs, tf.float32)
+        return ((max - min) / 255.) * obs + min
+
+    def _traverse_path(obs, path):
+        """Traverse `path` and transform the image at the path end"""
+        if not path:
+            return _transform_image(obs)
+        step = path[0]
+        return obs._replace(
```

Comment:
obs is not always a namedtuple. It can be dictionary, list, unnamed tuple, namedtuple. I think @Haichao-Zhang might use unnamed tuple in his task.

Reply:
If someone is going to this transformer and obs is a nest structure, it has to be a namedtuple; otherwise there is simply no way to specify which field to transform. Requirement has been written in the comment.

Reply:
Added support for dictionary

Reply:
I'm using named tuple as the observation structure. Unnamed tuple is not allowed by TFPyEnvironment.